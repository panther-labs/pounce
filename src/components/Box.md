Responsive box-model layout component.

```jsx harmony
import Text from 'components/Text';

<Box p={9} fontSize={4} width={300} color="white" bg="primary200">
  <Text size="medium" color="white" fontSize={4}>
    Box
  </Text>
</Box>;
```
