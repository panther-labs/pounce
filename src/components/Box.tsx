import React from 'react';
import styled, {
  CSSObject,
  FlattenSimpleInterpolation,
  FlattenInterpolation,
} from 'styled-components';
import * as StyledSystem from 'styled-system';
import { Theme } from 'styled-system';

export interface BoxProps
  extends StyledSystem.SpaceProps,
    StyledSystem.ColorProps,
    StyledSystem.FontFamilyProps,
    StyledSystem.FontSizeProps,
    StyledSystem.FontWeightProps,
    StyledSystem.WidthProps,
    StyledSystem.MinWidthProps,
    StyledSystem.HeightProps,
    StyledSystem.GridRowProps,
    StyledSystem.GridColumnProps,
    StyledSystem.ShadowProps,
    StyledSystem.BorderProps,
    StyledSystem.PositionProps,
    StyledSystem.FlexProps {
  /** The native HTML element to use for this component.
   * @default "div"
   * */
  as?: React.ElementType;

  /** Additional custom inline CSS to pass to the element
   * @default "{}"
   * */
  css?: CSSObject | FlattenSimpleInterpolation | FlattenInterpolation<Theme> | string;
}

/** Responsive box-model layout component. Apart from the defined props,
 * it also supports all the native HTML attributes. */
const Box: React.FC<BoxProps> = styled.div`
  ${StyledSystem.space}
  ${StyledSystem.color}
  ${StyledSystem.fontFamily}
  ${StyledSystem.fontSize}
  ${StyledSystem.fontWeight}
  ${StyledSystem.width}
  ${StyledSystem.minWidth}
  ${StyledSystem.height}
  ${StyledSystem.gridRow}
  ${StyledSystem.gridColumn}
  ${StyledSystem.shadow}
  ${StyledSystem.border}
  ${StyledSystem.position}
  ${StyledSystem.flex}
`;

export default Box;
