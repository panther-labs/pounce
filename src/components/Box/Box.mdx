Responsive box-model layout component.

The Box component is the primary component that powers the entire library. It accepts as props
almost all CSS properties and automatically converts theme values to actual browser-friendly values.

For example you can say `fontSize="medium"` and it will know to translate it to proper rem/px values.

A Box with the 9th value on the spacing scale, with large font and blue background:

```typescript jsx
<Box p={9} width={300} bg="blue-400" fontSize="large">
  Box
</Box>
```

A Box with the red font color and huge font-size:

```typescript jsx
<Box color="red-300" fontSize="2x-large">
  Box
</Box>
```

A Box that has a responsive width according to the screen size:

```typescript jsx
<Box p={9} bg="red-200" width={[300, 600, 900]} fontSize="2x-large">
  Box
</Box>
```

A Box can also implement other components under the hood and accept its props:

```typescript jsx
<Box as={Button} variantColor="pink-700">
  Box
</Box>
```
