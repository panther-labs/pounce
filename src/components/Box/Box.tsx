import React from 'react';
import { pounce, NativeAttributes, PounceComponentProps } from 'system';
import { __DEV__ } from 'utils/helpers';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type BoxProps<T extends React.ElementType = any> = PounceComponentProps<T>;
export type { NativeAttributes };

/**
 * Responsive box-model layout component. Apart from the defined props,
 * it also supports all the native HTML attributes.
 * */
const Box = pounce('div');

// FIXME: This overrides the fact that the box is a `div`. Components implementing this will warn us
//  that they can't implement `<Box>` when they themselves are not a div. We should ideally fix that
//  and have `ref` can be properly typed  in all components
export default Box as React.FC<BoxProps>;

if (__DEV__) {
  Box.displayName = 'Box';
}
