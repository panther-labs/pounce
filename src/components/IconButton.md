```jsx harmony
import Arrow from 'icons/arrow.svg';

<IconButton>
  <Arrow />
</IconButton>;
```
