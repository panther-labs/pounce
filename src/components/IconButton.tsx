import React from 'react';
import { css } from '@emotion/core';
import { convertHexToRgba } from 'utils/helpers';
import BaseButton, { BaseButtonProps } from 'components/BaseButton';
import useTheme from 'utils/useTheme';

export interface IconButtonProps extends BaseButtonProps {
  /** Whether the component should appear as "activated". Helpful for menus */
  active?: boolean;

  /** The color variant of the IconButton */
  variant: 'primary' | 'default';
}

/** A wrapper that makes an <a href="/#/Icon">Icon<a> component be clickable */
export const IconButton: React.FC<IconButtonProps> = ({ active, variant, ...rest }) => {
  const { colors } = useTheme();

  return (
    <BaseButton
      {...rest}
      bg="transparent"
      p={3}
      borderRadius="circle"
      css={css`
        color: ${active
          ? colors[variant === 'primary' ? 'primary300' : 'grey400']
          : colors.grey400};

        background-color: ${active &&
          convertHexToRgba(colors[variant === 'primary' ? 'primary300' : 'black'], 0.2)};

        &:hover:not(:active) {
          background-color: ${!active &&
            convertHexToRgba(colors[variant === 'primary' ? 'primary300' : 'black'], 0.1)};
        }

        &:active {
          background-color: ${convertHexToRgba(
            colors[variant === 'primary' ? 'primary300' : 'black'],
            0.2
          )};
        }
      `}
    />
  );
};

IconButton.defaultProps = {
  active: false,
};

export default IconButton;
