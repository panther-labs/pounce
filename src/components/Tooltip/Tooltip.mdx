The trigger of the tooltip can be anything. For example, an IconButton:

```typescript jsx
<Tooltip
  content={
    <Box>
      Lorem ipsum dolor sit amet, consectetur <br />
      adip iscing elit viverra <br />
      adip iscing elit viverra <br />
      adip iscing elit viverra <br />
      adip iscing elit viverra <br />
      adip iscing elit viverra <br />
      adip iscing elit viverra <br />
      adip iscing elit viverra <br />
      adip iscing elit viverra <br />
      neque nec est portt itor <br />
      dolor sit amet, consectetur.
    </Box>
  }
>
  <IconButton icon="user" aria-label="See user" />
</Tooltip>
```

Content can be anything as well:

```typescript jsx
<Tooltip content={<Icon size="small" type="user" />}>
  <Box>Hover me to see a pretty icon</Box>
</Tooltip>
```

The default right-side positioning can be customized:

```typescript jsx
// Center the tooltip, but collisions will win
const centered = (triggerRect, tooltipRect) => {
  const triggerCenter = triggerRect.left + triggerRect.width / 2;
  const left = triggerCenter - tooltipRect.width / 2;
  const maxLeft = window.innerWidth - tooltipRect.width - 2;
  return {
    left: Math.min(Math.max(2, left), maxLeft) + window.scrollX,
    top: triggerRect.bottom + 8 + window.scrollY,
  };
};

<Tooltip content="hello!" position={centered}>
  <Box>Hover me for a greeting</Box>
</Tooltip>;
```
