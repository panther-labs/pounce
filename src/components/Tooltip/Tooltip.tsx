import React from 'react';
import { useTooltip, TooltipPopup } from '@reach/tooltip';
import { Position } from '@reach/popover';
import { useTransition, animated } from 'react-spring';
import { positionRight } from './utils';
import Box from '../Box';

const AnimatedTooltipPopup = animated(TooltipPopup);

export interface TooltipProps {
  /** The string or HTML that the tooltip will show*/
  content: string | React.ReactElement;

  /** The position of the tooltip relative to the content */
  position: Position;

  /** @ignore */
  children: React.ReactElement;
}

/** A tooltip is a helper that shows some helping text when hovering or clicking something */
const Tooltip = React.forwardRef<HTMLDivElement, TooltipProps>(function Tooltip(
  { content, position = positionRight, children },
  ref
) {
  const [trigger, { triggerRect }, isVisible] = useTooltip();
  const transitions = useTransition(isVisible, null, {
    from: { opacity: 0, transform: 'scale(0.95, 0.95)' },
    enter: { opacity: 1, transform: 'scale(1, 1)' },
    leave: { opacity: 0, transform: 'scale(0.95, 0.95)' },
    config: { duration: 150 },
  });

  const childrenWithHandler = React.useMemo(() => React.cloneElement(children, trigger), [
    children,
    trigger,
  ]);

  return (
    <React.Fragment>
      {childrenWithHandler}
      {transitions.map(
        ({ item, key, props: styles }) =>
          item && (
            <AnimatedTooltipPopup
              triggerRect={triggerRect}
              isVisible={item}
              key={key}
              style={{ ...styles, zIndex: 11, position: 'absolute' }}
              ref={ref}
              position={position}
              as={'div'}
              label={
                <Box
                  borderRadius="medium"
                  bg="navyblue-300"
                  p={4}
                  m={2}
                  fontSize="small"
                  boxShadow="dark250"
                >
                  {content}
                </Box>
              }
            />
          )
      )}
    </React.Fragment>
  );
});

export default React.memo(Tooltip);
