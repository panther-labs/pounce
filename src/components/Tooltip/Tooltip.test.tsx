import React from 'react';
import { fireClickAndMouseEvents, fireEvent, renderWithTheme } from 'test-utils';
import Tooltip from './Tooltip';
import Box from '../Box';
import IconButton from '../IconButton';

const TooltipWrapper: React.FC = ({ children }) => {
  return (
    <Box
      position="relative"
      borderRadius="medium"
      bg="black"
      p={2}
      m={2}
      fontSize="small"
      boxShadow="dark250"
      __before={{
        content: '""',
        position: 'absolute',
        right: '100%',
        top: '50%',
        borderWidth: '6px',
        borderStyle: 'solid',
        borderTopColor: 'transparent',
        borderRightColor: 'black',
        borderBottomColor: 'transparent',
        borderLeftColor: 'transparent',
        transform: 'translateY(-50%)',
      }}
    >
      {children}
    </Box>
  );
};

describe('Tooltip', () => {
  it('matches snapshot for tooltip', () => {
    const { container } = renderWithTheme(
      <Tooltip content="Lorem ipsum dolor sit amet, consectetur">
        <Box>Hover me to see the tooltip</Box>
      </Tooltip>
    );

    expect(container).toMatchSnapshot();
  });

  it('shows the tooltip when we hover', async () => {
    const { getByText, findByText } = renderWithTheme(
      <Tooltip content="Boom! I am the tooltip">
        <Box>Hover me to see the tooltip</Box>
      </Tooltip>
    );

    const textToHover = getByText('Hover me to see the tooltip');
    expect(textToHover).toBeInTheDocument();
    fireClickAndMouseEvents(textToHover);
    fireEvent.mouseOver(textToHover);
    expect(await findByText('Boom! I am the tooltip')).toBeInTheDocument();
  });

  it('shows the tooltip with an Icon', async () => {
    const { getByLabelText, findByText } = renderWithTheme(
      <Tooltip content="Boom! I am the user!">
        <IconButton icon="user" aria-label="See user" />
      </Tooltip>
    );

    const textToHover = getByLabelText('See user');
    expect(textToHover).toBeInTheDocument();
    fireClickAndMouseEvents(textToHover);
    fireEvent.mouseOver(textToHover);
    expect(await findByText('Boom! I am the user!')).toBeInTheDocument();
  });

  it('shows the tooltip with a custom content', async () => {
    const { getByText, findByText } = renderWithTheme(
      <Tooltip content="Lorem ipsum dolor sit amet, consectetur" wrapper={TooltipWrapper}>
        <Box>Hover me to see the new wrapper</Box>
      </Tooltip>
    );

    const textToHover = getByText('Hover me to see the new wrapper');
    expect(textToHover).toBeInTheDocument();
    fireClickAndMouseEvents(textToHover);
    fireEvent.mouseOver(textToHover);
    const tooltip = await findByText('Lorem ipsum dolor sit amet, consectetur');
    expect(tooltip).toHaveStyle('background-color: rgb(0, 0, 0)');
  });
});
