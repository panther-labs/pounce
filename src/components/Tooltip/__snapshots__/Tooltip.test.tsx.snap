// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip matches snapshot for tooltip 1`] = `
<div>
  <div
    class="pounce-0"
    data-reach-tooltip-trigger=""
    data-state="tooltip-hidden"
  >
    Hover me to see the tooltip
  </div>
</div>
`;
